### PR TITLE
Non standard erc20 delimitation

### PR DESCRIPTION
**Pull request by chrismaree**
This pr contains an update to the unit tests to show how our system can accommodate non-standard erc20 delimitation within the EMP without needing to modify any contract code. Rather, this change occurs in how we deal with price feeds.

This fundamentally changes how we think about prices within the UMA ecosystem: **prices feeds now represent the value of synthetic tokens, represented in units of collateral currency.** 

Our current notion of a price of “_1 usd per synthetic token_” would be represented as 1e18, irrespective of the underlying scaling. In the new framing the price feed will be “_how many units of underlying make up one unit of synthetic?_” If the underlying is Dai and the synthetic is at 1 usd then the price feed would report 1e18. If the underlying is USDC then the price feed is scaled accordingly to this underlying as 1e6. 

This will require tooling to change upstream in anything that deals with a price. However I don't think this is too burdensome as this infrastructure would had needed to know about scaling prices in any event if non-18 delimited tokens are used.

The google doc outlining the different options available can be found [here](https://docs.google.com/document/d/1fQA6OgB5KMQz2SIN6xXUweXA9oY5a2eU9lOAteukm6Y/edit?usp=sharing).

### CODE REVIEW

**Comment by chrismaree** [comment]
> I generally agree with this framing, and I like the specifically targeted test.
> 
> I have two asks:
> 
> 1. Can you add a description to the PR describing this solution to the "USDC problem". Once you do that, I'd like to get both @ptare and @nicholaspai's eyes on this to see how they feel about pushing the burden to the offchain infra.
> 2. Our offchain infra is going to have to recognize and interpret token values relative to their decimalization. Can we just pull that from the `decimals()` function on all the tokens we care about that aren't 1e18 denominated? Or do some of them not define that method? If not, we may have to maintain our own list. Can you do a little digging on how common it is for tokens to implement the `decimals()` function?

Ok great. I will do 1. 

I've actually been thinking about 2 before. This makes our lives much easier. you can actually use the same mechanism within the contracts to scale things accordingly if you were to modify contract code however it makes more sense to leave the blockchain level stuff scaled according to it's native units and use this at the tooling level.

**Comment by chrismaree** [comment]
> So how does this affect the NFCT's dealing of the collateral token, specifically as a FixedPoint? I thought most of the work here would be in modifying FixedPoint to work with an arbitrary number of decimals?

This is how this task started; what you describe is in the google doc under solution 2. Originally I started modifying the `FixedPoint` lib to store the number of decimals associated with that fixed point variable and created means to convert into/out of different fixed point precision numbers. However it became apparent that the whole system would continue to operate fine without needing to change how FixedPoint works by simply modifying how we quantify price feeds. If we think about a price as notating the value of the synthetic in units of the underlying then we don't need to change any libs at all and everything operates in the same fashion. These tests show that this is the case for `PricelessPositionManager`.

**Comment by nicholaspai** [comment]
> I generally agree with this framing, and I like the specifically targeted test.
> 
> I have two asks:
> 
> 1. Can you add a description to the PR describing this solution to the "USDC problem". Once you do that, I'd like to get both @ptare and @nicholaspai's eyes on this to see how they feel about pushing the burden to the offchain infra.
> 2. Our offchain infra is going to have to recognize and interpret token values relative to their decimalization. Can we just pull that from the `decimals()` function on all the tokens we care about that aren't 1e18 denominated? Or do some of them not define that method? If not, we may have to maintain our own list. Can you do a little digging on how common it is for tokens to implement the `decimals()` function?

Totally agreed we should use `decimals()` if available, but I do worry that since this is non standard ERC20 not all contracts will have it. So maybe we will need to store not only the collateral token addresses but their decimals as well.

Overall, I think this is a pretty clever approach to the solution.

File: core/contracts/common/implementation/FixedPoint.sol [hunk 1]
```diff
@@ -17,7 +17,7 @@ library FixedPoint {
         uint rawValue;
     }
 
-    /** @dev Constructs an `Unsigned` from an unscaled uint, e.g., `b=5` gets stored internally as `5**18`. */
+    /** @dev Constructs an `Unsigned` from an unscaled uint, e.g., `a=5` gets stored internally as `5**18`. */
```
Comment (chrismaree) [inline core/contracts/common/implementation/FixedPoint.sol:20]: this is unrelated to this PR and is just a typo fix

File: core/contracts/common/implementation/TestnetERC20.sol [hunk 2]
```diff
@@ -9,10 +9,17 @@ import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
  * this token should never be used to store real value since it allows permissionless minting.
  */
 contract TestnetERC20 is ERC20 {
+    string public name;
```
Comment (chrismaree) [inline core/contracts/common/implementation/TestnetERC20.sol:12]: I had to modify this contract (and it's migration) to enable non-18 delimited tokens.

**Comment by chrismaree** [comment]
This has been moved to a new branch and PR. [see](https://github.com/UMAprotocol/protocol/pull/1159)
